File: backend/models/diagram.py
"""Pydantic models for diagram node/edge representation."""

from typing import Literal

from pydantic import BaseModel

from .scan import RegionFailure


class DiagramNode(BaseModel):
    """A node in the infrastructure diagram (represents one resource)."""

    id: str  # ARN
    resource_type: str
    name: str
    region: str
    is_external: bool = False
    is_unresolved: bool = False
    icon_url: str  # /api/images/icons/{service_type}


class DiagramEdge(BaseModel):
    """An edge in the infrastructure diagram (represents one relationship)."""

    id: str  # source_arn + target_arn hash
    source: str  # source ARN
    target: str  # target ARN
    category: Literal["network", "iam", "event", "data"]
    derived_from: str
    label: str | None = None


class DiagramData(BaseModel):
    """Complete diagram payload sent to the frontend."""

    nodes: list[DiagramNode]
    edges: list[DiagramEdge]
    account_id: str
    scan_timestamp: str
    total_resources: int
    scanned_regions: list[str]
    failures: list[RegionFailure] = []

File: frontend/src/types/diagram.ts
import type { RegionFailure } from "./resources";
import type { HierarchyTree } from "./hierarchy";

export interface DiagramNode {
  id: string;
  resource_type: string;
  name: string;
  region: string;
  is_external: boolean;
  is_unresolved: boolean;
  icon_url: string;
}

export interface DiagramEdge {
  id: string;
  source: string;
  target: string;
  category: "network" | "iam" | "event" | "data";
  derived_from: string;
  label: string | null;
}

export interface DiagramData {
  nodes: DiagramNode[];
  edges: DiagramEdge[];
  account_id: string;
  scan_timestamp: string;
  total_resources: number;
  scanned_regions: string[];
  failures: RegionFailure[];
  hierarchy: HierarchyTree | null;
}
